File: tictactoe.cs
using System;

class TicTacToe
{
    static char[][] InitBoard()
    {
        char[][] board = new char[3][];
        int cell = '1';
        for (int i = 0; i < 3; i++)
        {
            board[i] = new char[3];
            for (int j = 0; j < 3; j++)
            {
                board[i][j] = (char)cell;
                cell++;
            }
        }
        return board;
    }

    static void PrintBoard(char[][] board)
    {
        Console.WriteLine();
        for (int i = 0; i < 3; i++)
        {
            Console.WriteLine($" {board[i][0]} | {board[i][1]} | {board[i][2]} ");
            if (i != 2)
            {
                Console.WriteLine("---+---+---");
            }
        }
        Console.WriteLine();
    }

    static bool CheckWin(char[][] board)
    {
        // rows and columns
        for (int i = 0; i < 3; i++)
        {
            if (board[i][0] == board[i][1] && board[i][1] == board[i][2])
                return true;
            if (board[0][i] == board[1][i] && board[1][i] == board[2][i])
                return true;
        }
        // diagonals
        if (board[0][0] == board[1][1] && board[1][1] == board[2][2])
            return true;
        if (board[0][2] == board[1][1] && board[1][1] == board[2][0])
            return true;
        return false;
    }

    static bool IsDraw(char[][] board)
    {
        for (int i = 0; i < 3; i++)
        {
            for (int j = 0; j < 3; j++)
            {
                if (board[i][j] >= '1' && board[i][j] <= '9')
                    return false;
            }
        }
        return true;
    }

    static bool MakeMove(char[][] board, char player, int choice)
    {
        char target = choice.ToString()[0];
        for (int i = 0; i < 3; i++)
        {
            for (int j = 0; j < 3; j++)
            {
                if (board[i][j] == target)
                {
                    board[i][j] = player;
                    return true;
                }
            }
        }
        return false;
    }

    static void Main()
    {
        char player = 'X';
        bool gameOver = false;
        char[][] board = InitBoard();
        Console.WriteLine("Tic Tac Toe (Player X vs Player O)");
        PrintBoard(board);

        while (!gameOver)
        {
            Console.Write($"Player {player}, enter a cell number (1-9): ");
            string input = Console.ReadLine();
            int choice;
            if (!int.TryParse(input, out choice))
            {
                Console.WriteLine("Invalid input. Exiting.");
                break;
            }

            if (choice < 1 || choice > 9)
            {
                Console.WriteLine("Invalid input. Exiting.");
                break;
            }

            if (!MakeMove(board, player, choice))
            {
                Console.WriteLine("Cell already taken. Try again.");
                continue;
            }

            PrintBoard(board);

            if (CheckWin(board))
            {
                Console.WriteLine($"Player {player} wins!");
                gameOver = true;
            }
            else if (IsDraw(board))
            {
                Console.WriteLine("It's a draw.");
                gameOver = true;
            }
            else
            {
                player = (player == 'X') ? 'O' : 'X';
            }
        }
    }
}

File: tictactoe.rs
use std::io::{self, Write};

struct TicTacToe {
    board: [[char; 3]; 3],
}

impl TicTacToe {
    fn new() -> Self {
        let mut board = [[' '; 3]; 3];
        let mut cell = '1';
        for i in 0..3 {
            for j in 0..3 {
                board[i][j] = cell;
                cell = ((cell as u8) + 1) as char;
            }
        }
        TicTacToe { board }
    }

    fn print_board(&self) {
        println!();
        for i in 0..3 {
            println!(" {} | {} | {} ", self.board[i][0], self.board[i][1], self.board[i][2]);
            if i != 2 {
                println!("---+---+---");
            }
        }
        println!();
    }

    fn check_win(&self) -> bool {
        // rows and columns
        for i in 0..3 {
            if self.board[i][0] == self.board[i][1] && self.board[i][1] == self.board[i][2] {
                return true;
            }
            if self.board[0][i] == self.board[1][i] && self.board[1][i] == self.board[2][i] {
                return true;
            }
        }
        // diagonals
        if self.board[0][0] == self.board[1][1] && self.board[1][1] == self.board[2][2] {
            return true;
        }
        if self.board[0][2] == self.board[1][1] && self.board[1][1] == self.board[2][0] {
            return true;
        }
        false
    }

    fn is_draw(&self) -> bool {
        for i in 0..3 {
            for j in 0..3 {
                if self.board[i][j] >= '1' && self.board[i][j] <= '9' {
                    return false;
                }
            }
        }
        true
    }

    fn make_move(&mut self, player: char, choice: u32) -> bool {
        let target = std::char::from_digit(choice, 10).unwrap_or('0');
        for i in 0..3 {
            for j in 0..3 {
                if self.board[i][j] == target {
                    self.board[i][j] = player;
                    return true;
                }
            }
        }
        false
    }
}

fn main() {
    let mut game = TicTacToe::new();
    let mut player = 'X';
    let mut game_over = false;

    println!("Tic Tac Toe (Player X vs Player O)");
    game.print_board();

    while !game_over {
        print!("Player {}, enter a cell number (1-9): ", player);
        io::stdout().flush().unwrap();

        let mut input = String::new();
        if io::stdin().read_line(&mut input).is_err() {
            println!("Invalid input. Exiting.");
            break;
        }

        let choice: u32 = match input.trim().parse() {
            Ok(num) if num >= 1 && num <= 9 => num,
            _ => {
                println!("Invalid input. Exiting.");
                break;
            }
        };

        if !game.make_move(player, choice) {
            println!("Cell already taken. Try again.");
            continue;
        }

        game.print_board();

        if game.check_win() {
            println!("Player {} wins!", player);
            game_over = true;
        } else if game.is_draw() {
            println!("It's a draw.");
            game_over = true;
        } else {
            player = if player == 'X' { 'O' } else { 'X' };
        }
    }
}
